### PR TITLE
[curl] Update to 7.81.0

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?  
  Updates curl.
  Fixes `find_package(CURL COMPONENTS ...)` (cherry-picked from #22258).
  Drops the patch which disabled exporting the curl executable.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  unchanged, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

### CODE REVIEW

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
When building `curl[brotli,c-ares,core,http2,idn,idn2,mbedtls,non-http,openssl,schannel,ssh,ssl,sspi,tool,winidn,winldap,winssl]:x64-windows-static`:
```
FAILED: lib/CMakeFiles/libcurl.dir/vtls/openssl.c.obj 
"C:\PROGRA~2\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\Hostx64\x64\cl.exe"   -DBUILDING_LIBCURL -DCURLDEBUG -DCURL_HIDDEN_SYMBOLS -DDEBUGBUILD -DHAVE_CONFIG_H -DOPENSSL_SUPPRESS_DEPRECATED -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_SECURE_NO_DEPRECATE -D_WIN32_WINNT=0x0600 -IG:\22445\vcpkg\buildtrees\curl\src\url-7_81_0-7024fe0324.clean\include -IG:\22445\vcpkg\buildtrees\curl\x64-windows-static-dbg\lib\..\include -IG:\22445\vcpkg\buildtrees\curl\src\url-7_81_0-7024fe0324.clean\lib\.. -IG:\22445\vcpkg\buildtrees\curl\src\url-7_81_0-7024fe0324.clean\lib\..\include -IG:\22445\vcpkg\buildtrees\curl\x64-windows-static-dbg\lib\.. -IG:\22445\vcpkg\buildtrees\curl\src\url-7_81_0-7024fe0324.clean\lib -IG:\22445\vcpkg\buildtrees\curl\x64-windows-static-dbg\lib -IG:\22445\vcpkg\installed\x64-windows-static\include /nologo /DWIN32 /D_WINDOWS /W4 /utf-8 /MP  /MP /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1  -MTd /showIncludes /Folib\CMakeFiles\libcurl.dir\vtls\openssl.c.obj /Fdlib\CMakeFiles\libcurl.dir\libcurl.pdb /FS -c G:\22445\vcpkg\buildtrees\curl\src\url-7_81_0-7024fe0324.clean\lib\vtls\openssl.c
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(81): error C2059: syntax error: '<parameter-list>'
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(151): error C2059: syntax error: '<parameter-list>'
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(152): error C2059: syntax error: '('
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(152): error C2143: syntax error: missing ')' before '('
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(152): error C2059: syntax error: ')'
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(152): error C2143: syntax error: missing ')' before 'constant'
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(152): error C2091: function returns function
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(152): error C2143: syntax error: missing '{' before 'constant'
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(152): error C2059: syntax error: 'constant'
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(153): error C2059: syntax error: ')'
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(155): error C2143: syntax error: missing ')' before '('
G:\22445\vcpkg\installed\x64-windows-static\include\openssl/ocsp.h(155): error C2091: function returns function
```
Does that related?

**Comment by dg0yt** [comment]
@JackBoosY Maybe you need https://github.com/curl/curl/pull/8246 "To fix build errors when both Schannel and OpenSSL are enabled." Can you please test that change? (You know, I'm limited to vcpkg CI for windows triplets.)

**Comment by JackBoosY** [comment]
@dg0yt That's doesn't solve this, I checked.

**Comment by dg0yt** [comment]
> When building `curl[brotli,c-ares,core,http2,idn,idn2,mbedtls,non-http,openssl,schannel,ssh,ssl,sspi,tool,winidn,winldap,winssl]:x64-windows-static`:
> ~~~
> FAILED: lib/CMakeFiles/libcurl.dir/vtls/openssl.c.obj 
> ~~~

Not reproduced [in CI](https://dev.azure.com/vcpkg/public/_build/results?buildId=65655&view=results) with curl/curl#8246 applied.
However, there are issues with consuming the pc file.

**Comment by JackBoosY** [comment]
> > When building `curl[brotli,c-ares,core,http2,idn,idn2,mbedtls,non-http,openssl,schannel,ssh,ssl,sspi,tool,winidn,winldap,winssl]:x64-windows-static`:
> > ```
> > FAILED: lib/CMakeFiles/libcurl.dir/vtls/openssl.c.obj 
> > ```
> 
> Not reproduced [in CI](https://dev.azure.com/vcpkg/public/_build/results?buildId=65655&view=results) with [curl/curl#8246](https://github.com/curl/curl/pull/8246) applied. However, there are issues with consuming the pc file.

If this is no related to your changes, I can temporary skip this feature 😄

**Comment by dg0yt** [comment]
> If this is no related to your changes, I can temporary skip this feature

You should rather check if there was the same error with port curl 7.80.0. If yes, it is not a regression.

**Comment by dg0yt** [comment]
@JackBoosY The x64-windows-static debug build of curl now tries to link to pthreadVC3.lib (and fails) when it should use pthreadVC3d.lib. Some notes.

The given feature combination pulls in mbedtls.

mbedtls comes with cmake wrapper, although a standard mbedtls Find module doesn't exist in CMake. (Curl has one.)
mbedtls also pulls in the pthreads port for windows. 

pthreads comes with another cmake wrapper, which is installed for pthread, pthreads and pthread_windows package names. None of these names have a standard Find module in CMake. (mbedtls calls `find_package(pthreads_windows QUIET)`.)

The wrappers from mbedtls and pthreads don't wrap anything but are implemented like Find modules, including `find_package_handle_standard_args`. Which is a bad idea because wrapper don't see the `find_package` interface variables.

I still don't see why curl pulls the wrong lib but IMO pthreads and mbedtls are already broken.

**Comment by Neumann-A** [comment]
> I still don't see why curl pulls the wrong lib but IMO pthreads and mbedtls are already broken.

I tried your branch and doing `./vcpkg.exe install mbedtls[pthreads]:x64-windows-static curl:x64-windows-static` does not pull in the wrong pthread library.

**Comment by Neumann-A** [comment]
again a case for 
[cmake/#22621](https://gitlab.kitware.com/cmake/cmake/-/issues/22621)
maybe add a comment ;)

**Comment by Neumann-A** [comment]
@dg0yt: The debug pkg-config file still has `-ldebug -lnormaliz` while the release version has: `-lx64-windows-static -lnormaliz`

**Comment by JackBoosY** [comment]
Should I check something?

**Comment by dg0yt** [comment]
> Should I check something?

As I said, it would be good to know the if that feature combination works for 7.80.0. If not, we should prioritize the update over fixing the combination. If you don't test locally, I may need to create another draft PR.

Anyway, I try to solve the problems. The pthreadVC3[d].lib issue is in downstream usage via libcurl.pc. My position still stands, the mbedtls wrapper implementation - only installed with feature threads - is a ticking bomb. Googling for FindMbedTLS gives alternative find modules with different variables. Hard to refactor without `vcpkg_cmake_validate`. And of course mbedtls doesn't build for mingw without changes, and it is outdated...

**Comment by dg0yt** [comment]
> again a case for
> [cmake/#22621](https://gitlab.kitware.com/cmake/cmake/-/issues/22621)
> maybe add a comment ;)

@Neumann-A I recognized the pattern, too. With more features enabled here, the problem gets bigger:
- Port libssh2 always enables zlib.
- Curl's FindLibSSH2.cmake doesn't know about the zlib dependency. Now I realized that this was the reason for the strange zlib patch I just removed: When building the curl tool, libssh2 needs zlib after it in static linkage on linux. Curl adds zlib when it is enabled, but inserts it before libss2.
- Using LibSSH2 cmake config solves the problem in building curl.
  However, the libssh2 config pulls it is dependencies via `IMPORTED_LINK_INTERFACE_LIBRARIES_<Config>` which is not handled in the curl build when collecting libs for pc file and `curl-config` generation.
  So the export pc file has `-lz` only before `-lssh2`

It is really annoying...

**Comment by dg0yt** [comment]
Yet another comment:
Port curl lets the user choose SSL backends via features.
libssh2 supports different backends, too. However, the vcpkg port only supports openssl.
Ideally, if a user chose to build curl[core,mbedtls,ssh2], this would depend on libssh2[core,mbedtls]. However, there is no way to express dependencies of feature combinations :disappointed:

File: ports/curl/cmake-project-include.cmake [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Process the libs and targets in the variable named by `input`
+# into a flat list of libs in the variable named by `output`.
+# Simplify -framework elements.
+# Use -l where possible.
+# Avoid duplicates.
+function(vcpkg_curl_flatten input output)
```
Comment (dg0yt) [inline ports/curl/cmake-project-include.cmake:6]: @Neumann-A This is the updated version of transfering targets into link libraries for `*-config`. Dealing with a fairly long list of dependencies here when all features are enabled. Still not hitting generator expressions, and ignoring compile flags etc.

**Comment by JackBoosY** [comment]
Depends on #22534.

**Comment by JackBoosY** [comment]
Tested all features successfully on `x86-windows` and `x64-windows-static`.

**Pull request merged**